app/models/key.rb

```diff
@@ -1,4 +1,85 @@
+require 'securerandom'
+
 class Key < ActiveRecord::Base
+  has_many :logs
   belongs_to :place
-  has_and_belongs_to_many :clients
+  belongs_to :guest
+  before_create :set_secret_url
+
+  def self.get_available_key keys
+    keys.each do |key|
+      client_key = key.client_keys.find_by(requested: true, used_at: nil)
+      master_key = key.client_keys.find_by(requested: true, unlimited_access: true)
+      return client_key if client_key
```

Comment:
This needs re-ordering maybe. There's no point loading the master key if the client key was found.